docs/source/rfc-configuration-overhaul.md

```diff
@@ -0,0 +1,893 @@
+```
+
+This is not a helpful error message as it does not mention what exactly is wrong, why and where the parsing failed.
+
+#### 7. Invalid `IROHA2_PUBLIC_KEY` environment variable
+
+Providing an invalid `IROHA2_PUBLIC_KEY` environment variable results in the following error:
+
+```
+Error:
+   0: Failed to build configuration from env
+   1: Failed to deserialize the field `IROHA_PUBLIC_KEY`: JSON5: Key could not be parsed. Key could not be parsed. Invalid character 's' at position 4
+   2: JSON5: Key could not be parsed. Key could not be parsed. Invalid character 's' at position 4
+```
+
+While this message is more helful than the ones we discussed above, there are still issues:
+
+- The message contains repetition.
+- Without an input snippet, the `Invalid character 's' at position 4` part of the message is not as helpful as it could
+  be.
+
+#### 8. Config file with extra fields
+
+Providing a configuration file with extra fields that are not supposed to be there does not produce any errors at all.
+
+This might lead to a bad user experience when user expects some options to apply, but doesn't have any idea that those
+in fact are silently ignored.
+
+#### 9. Specifying `SUMERAGI.PUBLIC_KEY`
+
+As you can see in the "Redundant fields" section, user should not specify any non-null value for `SUMERAGI.PUBLIC_KEY`
+parameter. However, if they do, they will get the following error:
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+The problem is that the error message tells something completely unrelated to the actual cause. This produces a terrible
+experience for users.
+
+#### 10. Config file is an invalid JSON file
+
+Providing an invalid JSON file as a config results in the same error we've seen before, which is not useful at all in
+this particular case and doesn't tell the user anything about the invalid JSON:
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+### Ambiguity in Setting Root-Level Fields via ENV
+
+To understand the issue with ambiguity at the root-level fields, let's once again look at the excerpt from the
+[configuration reference](https://github.com/hyperledger/iroha/blob/35ba182e1d1b6b594712cf63bf448a2edefcf2cd/docs/source/references/config.md#network):
+
+> **`network`**
+>
+> Network configuration
+>
+> Has type `Option<network::ConfigurationProxy>`. Can be configured via environment variable `IROHA_NETWORK`
+>
+> ```json
+> {
+>   "ACTOR_CHANNEL_CAPACITY": 100
+> }
+> ```
+
+As we can see, `network` is not actually a configuration field in itself. Instead, there are various `network.*` fields
+that can be configured. This bring a question of what will happen when we set the `IROHA_NETWORK` environment variable
+as mentioned in the excerpt above? Will it override the nested `IROHA_NETWORK_CAPACITY` fields? Is there a use case for
+providing an ability to set `IROHA_NETWORK` all-in-one through an environment variable?
```

Comment:
`IROHA_NETWORK` is just a prefix/level, not the parameter, so should be treated as an excessive value

Reply:
_Currently_, due to implementation details, one can pass `IROHA_NETWORK` env var and it will work:

```bash
# set a particular field
IROHA_NETWORK_CAPACITY=100 ./iroha

# same
IROHA_NETWORK="{\"ACTOR_CHANNEL_CAPACITY\":100}" ./iroha
```

> **Note:** it is relevant not only to `network.*`, but to all other nestings, e.g. `torii.*`, `sumeragi.*` etc.

The problem now is that `network.*` treated not as a namespace of parameters, but also a parameter itself. I don't think we need such a feature. Moreover, although it is partially highlighted in the configuration reference, it is not well documented and produces questions I have highlighted in the RFC.

Will rephrase this part.

Reply:
That's kinda exotic, we'd rather exclude this kind of possibility imo

Reply:
Rephrased this bit, hope it is clearer now.